_posts/2019-01-04-nextjs.md

```diff
@@ -0,0 +1,290 @@
+};
+
+export default () => {
+    return <div>{hotel.name}</div>
+}
+```
+
+And then we can use the awesome `next/dynamic`:
+
+```
+import React from 'react';
+import dynamic from 'next/dynamic';
+
+const Hotel = dynamic(import('../src/hotel'));
+
+export default class extends React.Component {
+    state = {
+        showHotel: false
+    }
+
+    render() {
+        if (this.state.showHotel) {
+            return <Hotel />
+        }
+        return (<div>
+            <button onClick={() => this.setState({ showHotel: true })}>Show Hotel</button>
+        </div>)
+    }
+}
+```
+
+Do you see what changed? We're wrapping our dynamic import with `dynamic()`. This will create a react component that will dynamically import the needed code in the momement where we actually try to render the component.
+Sounds a bit weird at first ant the benefits are not immediately obvious. But `dynamic` has a second `options` param.
+
+And just changig it from:
+`const Hotel = dynamic(import('../src/hotel'));`
+
+to
+
+```
+const Hotel = dynamic(import('../src/hotel'),
+   {
+     loading: () => <p>LOADING...</p>
+   });
+```
+
+Will give us a "spinner" or loading screen for free. (Actually in newer react versions there is something comparable with React.Suspense - but that one only works on the client while next's also works with SSR).
+
+## Styling
+So how does styling work with next.js? Next bundles `styled-jsx` so you can style your application by embedding `<style jsx>` tags:
+
+```
+const hotel = {
+...
+};
+
+export default () => {
+    return (<div>
+        <style jsx>{`
+            div {
+                color: red;
+            }
+        `}</style>
+        {hotel.name}
+    </div>)
+}
+```
+While this looks alomst like using a normal style tag it has some benefits: It will only apply the styles to exactly this file. So no other `div` will get a red color.
+
+And if you don't like styled-jsx you can always switch to sass, inline-css, css-in-js with [https://github.com/zeit/next.js/#importing-css--sass--less--stylus-files](just a little configuration).
+
+## Wrapping up
+
+Next.js is a powerful tool for building single page application. As it follows convention over configuration principle, it makes it look easy to do this. Projects build with Next.js mostly have the same or very similar structure. Obvious use case where you definitely have to consider using Next.js would be building prototypes. It will be also handy for teams who don't want to maintain their own webpack config and setups, but still have advantages that framework provides.
+Next.js is by all means production ready and would fit for enterprise application and not only some small projects. It is built in a way that enables you to access lower level parts when needed, which makes it extremely flexible in what it does.
```

Comment:
I think it would be good to add a link to showcase gallery here, so the reads could see for themselves who and for what projects uses NextJS
https://nextjs.org/showcase/